### PR TITLE
DbTimestampedEntity

### PR DESCRIPTION
**Pull request by swankjesse**
This builds on our own event listeners.

### CODE REVIEW

File: misk/src/main/kotlin/misk/hibernate/HibernateModule.kt [hunk 1]
```diff
@@ -77,5 +78,16 @@ class HibernateModule(
...
+
+    install(object : HibernateEntityModule(qualifier) {
+      override fun configureHibernate() {
+        bindListener(EventType.PRE_INSERT).to(TimestampListener::class.java)
+        bindListener(EventType.PRE_UPDATE).to(TimestampListener::class.java)
+      }
+    })
   }
+
+  override fun equals(other: Any?) = other is HibernateModule && other.qualifier == qualifier
```
Comment (ryanhall07) [inline misk/src/main/kotlin/misk/hibernate/HibernateModule.kt]: this is to allow multiple install calls? why does that happen?

File: misk/src/main/kotlin/misk/hibernate/DbTimestampedEntity.kt [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+ *   override lateinit var created_at: Instant
+ *
+ *   ...
+ *
+ * }
+ * ```
+ *
+ * The timestamp listener will automatically populate these values on save and update.
+ */
+interface DbTimestampedEntity {
```
Comment (ryanhall07) [inline misk/src/main/kotlin/misk/hibernate/DbTimestampedEntity.kt]: do you think there is any value in splitting this interface up into mutable and immutable entities? I always found it unfortunate that an immutable entity was forced to have an `updated_at` field

File: misk/src/main/kotlin/misk/hibernate/HibernateModule.kt (hunk 1 above)
Comment (swankjesse) [reply misk/src/main/kotlin/misk/hibernate/HibernateModule.kt]: I added it when I was trying to diagnose why I was getting Hibernate errors. But I totally don’t need it. I’ll remove.

File: misk/src/main/kotlin/misk/hibernate/DbTimestampedEntity.kt (hunk 2 above)
Comment (swankjesse) [reply misk/src/main/kotlin/misk/hibernate/DbTimestampedEntity.kt]: Yeah, in this case I’m not quite ready to add that flexibility. In the spirit of being opinionated I want to have fewer types so our users have fewer decisions.

If you really need it, on immutable entities it’s very easy to set your own created_at date.

Comment (ryanhall07) [reply misk/src/main/kotlin/misk/hibernate/DbTimestampedEntity.kt]: yea good point!

File: misk/src/main/kotlin/misk/hibernate/DbTimestampedEntity.kt [hunk 3]
```diff
@@ -0,0 +1,54 @@
+package misk.hibernate
+
+import java.time.Instant
+
+/**
+ * This complements [DbEntity] for entities that keep updated_at and created_at timestamps.
+ *
+ * Misk uses the application's clock time to set [created_at] and [updated_at]. We prefer the
+ * application's (ie. JVM's) clock over the database's clock because only the application's clock
+ * can be faked in tests.
```
Comment (tirsen) [inline misk/src/main/kotlin/misk/hibernate/DbTimestampedEntity.kt]: This isn't necessarily true... We could potentially do something here at the Vitess level...

Comment (swankjesse) [reply misk/src/main/kotlin/misk/hibernate/DbTimestampedEntity.kt]: yeah, but that’s also pretty awkward to hook into FakeClock, right?

**Pull request merged**